### PR TITLE
fix: returning only completed/placed attendees

### LINKED ISSUE #6396: Return list of completed attendees instead of all
**Describe the bug**
While fetching the list of attendees using the endpoint `events/{id}/attendees` all attendees are returned while only completed attendees should be returned.

### PR DESCRIPTION
**Pull request by akshat0047**
Fixes #6396 

#### Short description of what this resolves:
Earlier /events/{event_id}/attendees was returning all the attendees even with order status pending

#### Changes proposed in this pull request:

- /events/{event_id}/attendees only returns completed/placed attendees
- attendees for an order_id, user_id and ticket_id also shows the same behaviour
- Made code efficient by removing implementing a condition for a function call

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by iamareebjamal** [comment]
We want admins to be able to see all and normal users to see just accepted. 

One way is to check the user and return the appropriate query, but more proper but involving thing is to create a separate endpoint for admins and using that in frontend admin section. That should have been the case from the beginning, but not sure how possible it will be to integrate now

**Comment by akshat0047** [comment]
Ok but do we need the accepted check for just events attendees or for orders/users/ticket id too?

**Comment by akshat0047** [comment]
Meanwhile I solve hound violations is it possible that we merge this one and open an issue to create an endpoint exclusively for admin?

**Comment by akshat0047** [comment]
There are two ways:

- we may merge this and open an issue for new end point, once created open an issue on frontend to move to the new endpoint and later removing the user check from this one.

- we may open a sub-issue under this then on frontend then merge the PR's respectively this will also hold/effect the issue #6591

**Comment by akshat0047** [comment]
Please suggest a solution @prateekj117 @iamareebjamal

**Comment by iamareebjamal** [comment]
We can't merge this and later add additional endpoint because it'll break frontend

**Comment by iamareebjamal** [comment]
We need only accepted sessions, attendees, etc. Orders should be only visible to the user itself. For example, if I say /events/orders, it should only return orders which I created. That too, only the ones which are pending or completed/placed.

**Comment by akshat0047** [comment]
Ok yes, sorry, frontend is still using the admin thing, may I open an issue and reference it under this for adding an endpoint for admin.

**Comment by iamareebjamal** [comment]
Yes

**Comment by akshat0047** [comment]
> We need only accepted sessions, attendees, etc. Orders should be only visible to the user itself. For example, if I say /events/orders, it should only return orders which I created. That too, only the ones which are pending or completed/placed.

but we are not considering /events/{event_id}/orders here instead /orders/{order_identifier}/attendees, this should be like for the new admin endpoint it will return order for pending/placed/completed for all the users. For non-admin i.e. present endpoint it will return pending/placed/completed orders for the current user only. Implementing the part with context to this issue.

**Comment by prateekj117** [comment]
@akshat0047 No need to make a new endpoint. Just check the user access level and return the response accordingly.

See https://github.com/fossasia/open-event-server/pull/6328.

In the above PR, I forgot to handle the Admin case. This solution is also similar to that along with the additional Admin check.

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/open-event-orga-server/open-event-orga-server/pullRequest?prid=4493609&bid=15153570)

File: app/api/attendees.py [hunk 1]
```diff
@@ -101,8 +102,14 @@ def query(self, view_kwargs):
...
-        query_ = self.session.query(TicketHolder)
-
+        if 'Authorization' in request.headers and not (has_access('is_admin') or has_access('is_super_admin')):
+            order_subq = self.session.query(Order).\
+                         filter(or_(Order.status == 'completed', Order.status == 'placed')).\
+                         subquery()
+            query_ = self.session.query(TicketHolder).join(order_subq, TicketHolder.order)
+        else:
+            query_ = self.session.query(TicketHolder).join(TicketHolder.order)
+
```
Comment (niranjan94) [inline app/api/attendees.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4493609)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/open-event-orga-server/open-event-orga-server/pullRequest?prid=4493609&bid=15153570)

File: app/api/attendees.py [hunk 2]
```diff
@@ -101,8 +102,15 @@ def query(self, view_kwargs):
...
-
+        
+        if "Authorization" in request.headers and not (has_access('is_admin') or has_access('is_super_admin')):
+            order_subq = self.session.query(Order).\
+                         filter(or_(Order.status == 'completed', Order.status == 'placed')).\
+                         subquery()
+            query_ = self.session.query(TicketHolder).join(order_subq, TicketHolder.order)
+        else:
+            query_ = self.session.query(TicketHolder)
+
```
Comment (niranjan94) [inline app/api/attendees.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/app/fossasia/open-event-orga-server/pullRequest?prid=4493609)

**Comment by akshat0047** [comment]
I've added checks for admin in my previous work, and its working the way it should, returning completed placed to others while all for admins/super admins. Can we merge this? @prateekj117  @iamareebjamal

File: app/api/attendees.py [hunk 3]
```diff
@@ -101,7 +102,13 @@ def query(self, view_kwargs):
         :param view_kwargs:
         :return:
         """
-        query_ = self.session.query(TicketHolder)
+
+        if "Authorization" in request.headers and not (has_access('is_admin') or has_access('is_super_admin')):
```
Comment (iamareebjamal) [inline app/api/attendees.py]: Looks good, but `"Authorization" in request.headers` is not needed elsewhere where same stuff is checked

File: app/api/attendees.py [hunk 4]
```diff
@@ -101,7 +102,13 @@ def query(self, view_kwargs):
...
         :return:
         """
-        query_ = self.session.query(TicketHolder)
+
+        if not (has_access('is_admin') or has_access('is_super_admin')):
+            order_subq = self.session.query(Order).\
+                         filter(or_(Order.status == 'completed', Order.status == 'placed')).subquery()
+            query_ = self.session.query(TicketHolder).join(order_subq, TicketHolder.order)
+        else:
+            query_ = self.session.query(TicketHolder)
```
Comment (iamareebjamal) [inline app/api/attendees.py]: `query_ = self.session.query(TicketHolder)`

This section is being repeated. Remove the duplication

File: app/api/attendees.py [hunk 5]
```diff
@@ -101,7 +102,13 @@ def query(self, view_kwargs):
         :param view_kwargs:
         :return:
         """
-        query_ = self.session.query(TicketHolder)
+
```
[the remark this answers is not in the document]
Comment (iamareebjamal) [reply app/api/attendees.py]: Please don't resolve hound comments unless they are outdated

**Comment by akshat0047** [comment]
Done

File: app/api/attendees.py [hunk 6]
```diff
@@ -122,7 +129,9 @@ def query(self, view_kwargs):
                 raise ForbiddenException({'source': ''}, 'Access Forbidden')
             query_ = query_.join(User, User.email == TicketHolder.email).filter(User.id == user.id)
 
-        query_ = event_query(self, query_, view_kwargs, permission='is_registrar')
+        if view_kwargs.get('event_id'):
```
Comment (iamareebjamal) [inline app/api/attendees.py]: Why is this needed?

Comment (akshat0047) [reply app/api/attendees.py]: Because we're running it irrespective of we have event_id in view_kwargs or not
Though its checked inside the function, but I made it so that we don't have to access the function, if not required
Screenshot_2019-11-24_14-34-15

Comment (iamareebjamal) [reply app/api/attendees.py]: It's unnecessary check and will break if the endpoint is accessed using event_identifier

File: app/api/attendees.py [hunk 7]
```diff
@@ -123,6 +130,7 @@ def query(self, view_kwargs):
             query_ = query_.join(User, User.email == TicketHolder.email).filter(User.id == user.id)
 
         query_ = event_query(self, query_, view_kwargs, permission='is_registrar')
+
```
Comment (iamareebjamal) [inline app/api/attendees.py]: Revert

File: app/api/attendees.py [hunk 8]
```diff
@@ -101,8 +102,14 @@ def query(self, view_kwargs):
         :param view_kwargs:
         :return:
         """
+
```
Comment (iamareebjamal) [inline app/api/attendees.py]: Revert

File: app/api/attendees.py (hunk 7 above)
Comment (akshat0047) [reply app/api/attendees.py]: But you said it's unnecessary check and will break the end-point ?

Comment (iamareebjamal) [reply app/api/attendees.py]: Revert the newline added

**Comment by iamareebjamal** [comment]
@kushthedude @prateekj117 Review

File: app/api/attendees.py [hunk 9]
```diff
@@ -103,6 +104,11 @@ def query(self, view_kwargs):
         """
         query_ = self.session.query(TicketHolder)
 
+        if not (has_access('is_admin') or has_access('is_super_admin')):
```
Comment (prateekj117) [inline app/api/attendees.py]: It would not show expired/initialized orders to the Owner/Organizer/Co-organizer in the event dashboard.

**Comment by iamareebjamal** [comment]
It has been solved. Only admins have access to attendees now
